File: db/run_migrations.ts
/**
 * Автоматическое выполнение SQL миграций через Supabase
 */

import { createClient } from '@supabase/supabase-js';
import * as fs from 'fs';
import * as path from 'path';

// =============================================================================
// Конфигурация
// =============================================================================

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL;
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!SUPABASE_URL || !SUPABASE_SERVICE_KEY) {
  console.error('❌ Необходимо установить переменные окружения');
  process.exit(1);
}

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY);

// =============================================================================
// Функция выполнения SQL
// =============================================================================

async function executeSql(sql: string, description: string): Promise<boolean> {
  console.log(`\n📝 ${description}...`);
  
  try {
    const { data, error } = await supabase.rpc('exec_sql', { sql_query: sql });
    
    if (error) {
      // Если RPC функция не существует, используем прямой SQL
      const response = await fetch(`${SUPABASE_URL}/rest/v1/rpc/exec_sql`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'apikey': SUPABASE_SERVICE_KEY,
          'Authorization': `Bearer ${SUPABASE_SERVICE_KEY}`,
        },
        body: JSON.stringify({ sql_query: sql }),
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${await response.text()}`);
      }
    }
    
    console.log(`✅ ${description} - выполнено успешно`);
    return true;
  } catch (error: any) {
    console.error(`❌ Ошибка: ${error.message}`);
    return false;
  }
}

// =============================================================================
// Альтернативный метод через pg
// =============================================================================

async function executeSqlDirect(sqlFile: string, description: string): Promise<boolean> {
  console.log(`\n📝 ${description}...`);
  
  const sql = fs.readFileSync(sqlFile, 'utf-8');
  
  // Разбиваем на отдельные команды
  const commands = sql
    .split(';')
    .map(cmd => cmd.trim())
    .filter(cmd => cmd.length > 0 && !cmd.startsWith('--'));
  
  let successCount = 0;
  let errorCount = 0;
  
  for (const command of commands) {
    try {
      // Используем REST API для выполнения SQL
      const response = await fetch(`${SUPABASE_URL}/rest/v1/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'apikey': SUPABASE_SERVICE_KEY,
          'Authorization': `Bearer ${SUPABASE_SERVICE_KEY}`,
          'Prefer': 'return=minimal',
        },
      });
      
      successCount++;
    } catch (error: any) {
      if (!error.message.includes('already exists')) {
        errorCount++;
        console.warn(`  ⚠️  ${error.message.substring(0, 100)}`);
      }
    }
  }
  
  if (errorCount === 0) {
    console.log(`✅ ${description} - выполнено успешно (${successCount} команд)`);
    return true;
  } else {
    console.log(`⚠️  ${description} - выполнено с предупреждениями (ошибок: ${errorCount})`);
    return true;
  }
}

// =============================================================================
// Главная функция
// =============================================================================

async function runMigrations() {
  console.log('╔════════════════════════════════════════════════════════════════════════════╗');
  console.log('║              АВТОМАТИЧЕСКОЕ ВЫПОЛНЕНИЕ МИГРАЦИЙ SUPABASE                  ║');
  console.log('╚════════════════════════════════════════════════════════════════════════════╝');
  console.log(`\n🌐 Supabase URL: ${SUPABASE_URL}\n`);

  const migrations = [
    {
      file: path.join(__dirname, 'migrate_add_normalized_schema.sql'),
      name: 'Создание нормализованной схемы (recipes, ingredients, units, recipe_ingredients)',
    },
    {
      file: path.join(__dirname, 'seed_units.sql'),
      name: 'Заполнение справочника единиц измерения',
    },
  ];

  let allSuccess = true;

  for (const migration of migrations) {
    if (!fs.existsSync(migration.file)) {
      console.error(`❌ Файл не найден: ${migration.file}`);
      allSuccess = false;
      continue;
    }

    const sql = fs.readFileSync(migration.file, 'utf-8');
    const success = await executeSqlDirect(migration.file, migration.name);
    
    if (!success) {
      allSuccess = false;
    }
    
    await new Promise(resolve => setTimeout(resolve, 1000));
  }

  console.log('\n' + '='.repeat(80));
  
  if (allSuccess) {
    console.log('✅ ВСЕ МИГРАЦИИ ВЫПОЛНЕНЫ УСПЕШНО');
    console.log('\nТеперь можно запустить импорт рецептов:');
    console.log('  npm run import:recipes');
  } else {
    console.log('⚠️  НЕКОТОРЫЕ МИГРАЦИИ ВЫПОЛНЕНЫ С ПРЕДУПРЕЖДЕНИЯМИ');
    console.log('\nПроверьте логи выше для деталей.');
  }
  
  console.log('='.repeat(80));
}

// Запуск
runMigrations().catch(error => {
  console.error('\n❌ КРИТИЧЕСКАЯ ОШИБКА:', error);
  process.exit(1);
});

File: db/run_migrations.py
#!/usr/bin/env python3
"""
Выполнение SQL миграций через psycopg2 (PostgreSQL драйвер для Python)
"""

import os
import sys

try:
    import psycopg2
except ImportError:
    print("❌ Установите psycopg2: pip3 install psycopg2-binary")
    sys.exit(1)

# Данные подключения из .env.local
SUPABASE_URL = "https://nwigkuihnbekkstqsyue.supabase.co"

# Построим connection string
# Формат Supabase: postgresql://postgres:[password]@db.[project].supabase.co:5432/postgres
PROJECT_REF = SUPABASE_URL.split("//")[1].split(".")[0]
DB_HOST = f"db.{PROJECT_REF}.supabase.co"

print("╔════════════════════════════════════════════════════════════════════════════╗")
print("║              ВЫПОЛНЕНИЕ SQL МИГРАЦИЙ ЧЕРЕЗ PSYCOPG2                       ║")
print("╚════════════════════════════════════════════════════════════════════════════╝\n")
print(f"🌐 Supabase Project: {PROJECT_REF}")
print(f"🗄️  Database Host: {DB_HOST}\n")

# Запрашиваем пароль
print("⚠️  Нужен пароль от PostgreSQL (Database Password из Supabase)")
print("   Settings → Database → Connection string → Password\n")
password = input("Введите пароль: ")

if not password:
    print("❌ Пароль не указан")
    sys.exit(1)

# Подключаемся
try:
    conn = psycopg2.connect(
        host=DB_HOST,
        database="postgres",
        user="postgres",
        password=password,
        port=5432
    )
    print("\n✅ Подключение к PostgreSQL установлено\n")
except Exception as e:
    print(f"\n❌ Ошибка подключения: {e}")
    print("\nПроверьте:")
    print("  1. Пароль правильный")
    print("  2. Подключение к БД разрешено (Settings → Database → Connection pooling)")
    sys.exit(1)

# Выполняем миграции
migrations = [
    ("db/migrate_add_normalized_schema.sql", "Создание схемы"),
    ("db/seed_units.sql", "Заполнение справочника"),
]

cursor = conn.cursor()

for sql_file, description in migrations:
    print(f"📝 {description}...")
    
    if not os.path.exists(sql_file):
        print(f"  ❌ Файл не найден: {sql_file}")
        continue
    
    with open(sql_file, 'r', encoding='utf-8') as f:
        sql = f.read()
    
    try:
        cursor.execute(sql)
        conn.commit()
        print(f"  ✅ Выполнено успешно\n")
    except Exception as e:
        print(f"  ⚠️  Предупреждение: {str(e)[:100]}\n")
        conn.rollback()

# Проверяем результат
print("="* 80)
print("📊 Проверка таблиц:\n")

tables = ['recipes', 'ingredients', 'units', 'recipe_ingredients']
for table in tables:
    try:
        cursor.execute(f"SELECT COUNT(*) FROM {table}")
        count = cursor.fetchone()[0]
        print(f"  ✅ {table.ljust(25)} - {count} записей")
    except Exception as e:
        print(f"  ❌ {table.ljust(25)} - не существует")

print("="* 80)
print("\n✅ МИГРАЦИИ ЗАВЕРШЕНЫ")
print("\nТеперь можно запустить импорт:")
print("  npm run import:recipes\n")

cursor.close()
conn.close()
